### PR TITLE
Tooltips & Utilities

### PR DESCRIPTION
**Pull request by supudo**
Me again :)

I've added the tooltip methods and most of the utilities methods.
If you have comments - docs or anything i missed would make changes gladly.

S.

### CODE REVIEW

File: imgui/core.pyx [hunk 1]
```diff
@@ -1374,6 +1374,68 @@ def is_window_collapsed():
     return cimgui.IsWindowCollapsed()
 
 
+def set_tooltip(str text):
+    """Set tooltip under mouse-cursor.
```
Comment (swistakm) [inline imgui/core.pyx]: Sorry for nitpicking but could you add additional new line after first row of docstring. Every other function is documented in style that are roughly compliant to PEP 257 style and it would be great to have this consistent.

BTW. it would be great to have this doctstring summary always fit exactly one line with 79 cols limit (this is actually a comment regarding other functions).

File: imgui/core.pyx [hunk 2]
```diff
@@ -1374,6 +1374,68 @@ def is_window_collapsed():
...
+    Usually used with :func:`is_item_hovered()`.
+    For a complex tooltip window see :func:`begin_tooltip()`.
+
+    .. visual-example::
+        :auto_layout:
+        :height: 100
+        :width: 200
+        :click: 10 10
```
Comment (swistakm) [inline imgui/core.pyx]: If you set click position to something like `80 40` we will be able to actually show the tooltip in the documentation example.

File: imgui/core.pyx [hunk 3]
```diff
@@ -1374,6 +1374,68 @@ def is_window_collapsed():
...
+
+
```
Comment (swistakm) [inline imgui/core.pyx]: redundant whitespace, single empty line is enough

File: imgui/core.pyx [hunk 4]
```diff
@@ -1374,6 +1374,68 @@ def is_window_collapsed():
...
+        imgui.begin("Example: tooltip")
+        imgui.button("Click me!")
```
Comment (swistakm) [inline imgui/core.pyx]: Could be `"Hover me!"` because we are displaying action on hovering, not clicking.

In future I will consider adding some `:hover:` option analogous to `:click:` for visual examples generation. Maybe we will even completely redesign how interactions are handled (animated examples?).

File: imgui/core.pyx [hunk 5]
```diff
@@ -1374,6 +1374,68 @@ def is_window_collapsed():
...
+        if imgui.is_item_hovered():
+            imgui.tooltip("Please?")
```
Comment (swistakm) [inline imgui/core.pyx]: Should be `imgui.set_tooltip()`.

File: imgui/core.pyx [hunk 6]
```diff
@@ -1374,6 +1374,68 @@ def is_window_collapsed():
...
+    # note: "%s" required for safety and to favor of Python string formating
+    cimgui.SetTooltip("%s", _bytes(text))
+
+
+def begin_tooltip():
+    """Use to create full-featured tooltip windows that aren't just text.
+
+    .. visual-example::
+        :auto_layout:
+        :height: 100
```
Comment (swistakm) [inline imgui/core.pyx]: You could resize the example canvas so whole tooltip will be displayed in docs. The:

        :height: 100
        :width: 200

should be fine

File: imgui/core.pyx [hunk 7]
```diff
@@ -1374,6 +1374,68 @@ def is_window_collapsed():
...
+        :width: 200
+        :click: 10 10
```
Comment (swistakm) [inline imgui/core.pyx]: This click could be changed too. `80 40` should be fine.

File: imgui/core.pyx [hunk 8]
```diff
@@ -1374,6 +1374,68 @@ def is_window_collapsed():
...
+
+
```
Comment (swistakm) [inline imgui/core.pyx]: redundant whitespace,  single empty line is enough :)

File: imgui/core.pyx [hunk 9]
```diff
@@ -1374,6 +1374,68 @@ def is_window_collapsed():
...
+        imgui.begin("Example: tooltip")
+        imgui.button("Click me!")
+        if imgui.is_item_hovered():
+            imgui.begin_tooltip()
+            imgui.text("This button is clickable.")
+            imgui.text("This button has full window tooltip.")
+            imgui.image(texture_id, 512, 64, border_color=(1, 0, 0, 1))
```
Comment (swistakm) [inline imgui/core.pyx]: This will not be rendered by our off-screen rendered because `texture_id` is not available in the local scope. You can get some example image using:

    texture_id = imgui.get_io().fonts.texture_id

See: `image()` function docstring

Comment (swistakm) [inline imgui/core.pyx]: Note that in future we could add some better example image textures for the documentation instead of relying on font atlas. Still, this is a very low priority.

File: imgui/core.pyx [hunk 10]
```diff
@@ -2638,7 +2723,241 @@ def is_item_hovered():
...
+
+    .. wraps::
+        bool IsItemClicked(int mouse_button = 0)
+    """
+    return cimgui.IsItemClicked(mouse_button)
+
+
+def is_item_visible():
+    """Was the last item visible?
+    (aka not out of sight due to clipping/scrolling.)
```
Comment (swistakm) [inline imgui/core.pyx]: Could you rephrase it so it will not require parentheses?

File: imgui/core.pyx [hunk 11]
```diff
@@ -2638,7 +2723,241 @@ def is_item_hovered():
...
+    to catch unused area.
+
+    .. wraps::
+        void SetItemAllowOverlap()
+    """
+    cimgui.SetItemAllowOverlap()
+
+
+def is_window_hovered():
+    """Is current window hovered and hoverable (not blocked by a popup)
```
Comment (swistakm) [inline imgui/core.pyx]: Missing dot at the sentence end.

File: imgui/core.pyx [hunk 12]
```diff
@@ -2638,7 +2723,241 @@ def is_item_hovered():
...
+        bool: True if any of the windows is hovered, otherwise False.
+
+    .. wraps::
+        bool IsRootWindowOrAnyChildHovered()
+    """
+    return cimgui.IsRootWindowOrAnyChildHovered()
+
+
+def is_rect_visible(float size_width, float size_height):
+    """Test if rectangle of given size starting from cursor pos
```
Comment (swistakm) [inline imgui/core.pyx]: Could you reformat this docstring to be more consistent with existing documentation style?

File: imgui/core.pyx [hunk 13]
```diff
@@ -2638,7 +2723,241 @@ def is_item_hovered():
...
+        bool: True if any of the windows is on focus, otherwise False.
+
+    .. wraps::
+        bool IsRootWindowOrAnyChildFocused()
+    """
+    return cimgui.IsRootWindowOrAnyChildFocused()
+
+
+def is_root_window_or_any_child_hovered():
+    """Is current root window or any of its child
```
Comment (swistakm) [inline imgui/core.pyx]: Could you reformat this docstring to be more consistent with existing documentation style?

File: imgui/core.pyx [hunk 14]
```diff
@@ -2638,7 +2723,241 @@ def is_item_hovered():
...
+        bool: True if root window is on focus, otherwise False.
+
+    .. wraps::
+        bool IsRootWindowFocused()
+    """
+    return cimgui.IsRootWindowFocused()
+
+
+def is_root_window_or_any_child_focused():
+    """Is current root window or any of its child
```
Comment (swistakm) [inline imgui/core.pyx]: Could you reformat this docstring to be more consistent with existing documentation style?

File: imgui/core.pyx [hunk 15]
```diff
@@ -1646,6 +1708,29 @@ def label_text(str label, str text):
     cimgui.LabelText(_bytes(label), "%s", _bytes(text))
 
 
+def text_unformatted(str text):
+    """Doesn't require ``NULL`` terminated string if 'text_end' is specified.
```
Comment (swistakm) [inline imgui/core.pyx]: Since we don't allow passing the `text_end` argument this docstring is not valid.

Comment (swistakm) [inline imgui/core.pyx]: Since we don't allow passing the `text_end` argument this docstring is not valid.

**Comment by supudo** [comment]
Tried to push directly in this branch again, might've broken something tho :)

Anyways, i'm trying to setup my docs system so that i can see the screenshots and the full docs generated. Still having some issues, but hopefully i'll get it running.

S.

**Comment by swistakm** [comment]
Don't worry! Pushing to the same branch is completely fine. Actually, it is a recommended practice because allows to reliably track new changes and review comments.
